File: market/test/java/ru/yandex/market/replenishment/autoorder/config/ControllerTest.java
package ru.yandex.market.replenishment.autoorder.config;

import java.io.IOException;
import java.io.InputStream;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Objects;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.SneakyThrows;
import org.apache.commons.io.IOUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;

import ru.yandex.market.replenishment.autoorder.security.WithMockLogin;

/**
 * Базовый класс для всех контроллеров. На нем не должно быть аннотаций {@link WithMockLogin}
 * или {@link ru.yandex.market.replenishment.autoorder.security.WithMockTvm}.
 * Так как аннотации конфликтуют друг с другом и разработчик должен выбирать какую аннотацию использовать.
 */
@AutoConfigureMockMvc
public abstract class ControllerTest extends FunctionalTest {
    @Autowired
    protected MockMvc mockMvc;
    @Autowired
    protected ObjectMapper objectMapper;

    @SneakyThrows
    public String readFile(String fileName) {
        InputStream inputStream = this.getClass().getResourceAsStream(fileName);
        return IOUtils.toString(Objects.requireNonNull(inputStream), StandardCharsets.UTF_8.name());
    }

    protected <T> T readJson(String content, Class<T> cls) throws IOException {
        return objectMapper.readValue(content, cls);
    }

    protected <T> T readJson(MvcResult content, Class<T> cls) throws IOException {
        return objectMapper.readValue(content.getResponse().getContentAsString(), cls);
    }

    protected <T> List<T> readJsonList(String content, Class<T> cls) throws IOException {
        return objectMapper.readValue(content, new TypeReference<List<T>>() {
            @Override
            public Type getType() {
                return new ParameterizedType() {
                    @Override
                    public Type[] getActualTypeArguments() {
                        return new Type[]{cls};
                    }

                    @Override
                    public Type getRawType() {
                        return List.class;
                    }

                    @Override
                    public Type getOwnerType() {
                        return null;
                    }
                };
            }
        });
    }

    protected <T> List<T> readJsonList(MvcResult content, Class<T> cls) throws IOException {
        return readJsonList(content.getResponse().getContentAsString(), cls);
    }

    protected <T> T readJson(String content, TypeReference<T> cls) throws IOException {
        return objectMapper.readValue(content, cls);
    }

    protected <T> T readJson(MvcResult result, TypeReference<T> cls) throws IOException {
        return readJson(result.getResponse().getContentAsString(), cls);
    }
}

File: market/test/java/ru/yandex/market/pricingmgmt/api/ControllerTest.kt
package ru.yandex.market.pricingmgmt.api

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.security.test.context.support.WithMockUser
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.MvcResult
import ru.yandex.market.pricingmgmt.AbstractFunctionalTest
import ru.yandex.market.pricingmgmt.config.security.passport.PassportAuthenticationFilter
import java.io.IOException

@AutoConfigureMockMvc
@WithMockUser(username = PassportAuthenticationFilter.LOCAL_DEV, roles = ["PRICING_MGMT_ACCESS"])
abstract class ControllerTest : AbstractFunctionalTest() {
    @Autowired
    protected lateinit var mockMvc: MockMvc

    @Autowired
    protected var objectMapper: ObjectMapper? = null

    @Throws(IOException::class)
    protected fun <T> readJson(content: String?, cls: Class<T>?): T {
        return objectMapper!!.readValue(content, cls)
    }

    @Throws(IOException::class)
    protected fun <T> readJson(content: MvcResult, cls: Class<T>?): T {
        return objectMapper!!.readValue(content.response.contentAsString, cls)
    }

    @Throws(IOException::class)
    protected fun <T> readJson(content: String?, cls: TypeReference<T>?): T {
        return objectMapper!!.readValue(content, cls)
    }

    @Throws(IOException::class)
    protected fun <T> readJson(result: MvcResult, cls: TypeReference<T>?): T {
        return readJson(result.response.contentAsString, cls)
    }

    @Throws(JsonProcessingException::class)
    protected fun dtoToString(dto: Any?): String {
        // objectMapper!!.configure(SerializationFeature.WRAP_ROOT_VALUE, false)
        val ow = objectMapper!!.writer().withDefaultPrettyPrinter()
        return ow.writeValueAsString(dto)
    }

    protected inline fun <reified T : Any> ObjectMapper.readValue(content: String): T =
        readValue(content, object : TypeReference<T>() {})
}
